packages/financial-templates-lib/src/price-feed/DefiPulseTotalPriceFeed.js

```diff
@@ -0,0 +1,146 @@
+      let past = this.historicalPrices[i].timestamp;
+      let val = this.historicalPrices[i].tvlUSD;
+
+      if (past > closestTime.timestamp && past < time) {
+        closestTime.timestamp = past;
+        closestTime.tvlUSD = val;
+      }
+    }
+
+    const historicalPrice = this.scaleResult(closestTime.tvlUSD);
+
+    if (closestTime.timestamp === 0) {
+      return undefined;
+    } else {
+      return historicalPrice;
+    }
+  }
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  async update() {
+    const currentTime = this.getTime();
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== undefined && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
+      this.logger.debug({
+        at: "DefiPulseTotalPriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTimes + this.minTimeBetweenUpdates - currentTime
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "DefiPulseTotalPriceFeed",
+      message: "Updating",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime
+    });
+
+    // 1. Construct URLs.
+    const priceUrl = `https://data-api.defipulse.com/api/v1/defipulse/api/GetHistory?period=1w&api-key=${this.apiKey}`;
+
+    // 2. Send requests.
+    const response = await this.networker.getJson(priceUrl);
+
+    // 3. Check responses.
+    if (!response) {
+      throw new Error(`🚨Could not parse price result from url ${priceUrl}: ${JSON.stringify(response)}`);
+    }
+
+    // 4. Parse results.
+    // Return data structure:
+    //        [{
+    //            "timestamp":"1611511200"
+    //            "tvlUSD":25583565042,
+    //             ...
+    //          },
+    //           {
+    //            "timestamp":"1611507600",
+    //            "tvlUSD":25177860561
+    //             ...
+    //          },
+    //        ]
+
+    // Get tvlUSD for most most recent timestamp
+    let mostRecent = { timestamp: 0, tvlUSD: 0 };
+    for (let i = 0; i < response.length; i++) {
+      if (Number(response[i].timestamp) > mostRecent.timestamp) {
+        mostRecent.timestamp = Number(response[i].timestamp);
+        mostRecent.tvlUSD = Number(response[i].tvlUSD);
+      }
+    }
+
+    const newPrice = this.scaleResult(mostRecent.tvlUSD);
+
+    // 5. Store results.
+    this.lastUpdateTime = currentTime;
+    this.currentPrice = newPrice;
+    this.historicalPrices = response;
+  }
+
+  scaleResult(_tvlUSD) {
+    // As described in UMIP 24
+    // In an effort to make the token price affordable, the value of the token is the tvlUSD divided by 1 billion.
+    // We also cut off precision after 3 decimals to match the specified price step of .001
+
+    const billion = 1_000_000_000;
+    return this.toWei((_tvlUSD / billion).toFixed(3), "ether");
```

Comment:
@chrismaree can correct me if I'm wrong, but I think you should make the `decimals` value (i.e. how many decimals you scale your output to) configurable with a constructor parameter.

You also need to add a `getPriceFeedDecimals()` method that returns the number of decimals that the price feed is outputting (which is a part of the PriceFeedInterface on master: https://github.com/UMAprotocol/protocol/blob/ec5908e93c72e547b94b2ccf675b3c5d94ad7b93/packages/financial-templates-lib/src/price-feed/PriceFeedInterface.js#L37).

Reply:
yup! it should be configurable with a constructor parameter that then gets defined in the `CreatePriceFeed` in the same way that the other feeds have their decimals fed in. See the code below for an example:
https://github.com/UMAprotocol/protocol/blob/c91e11bad264509efd4ef98044e448e6e5b8b5f0/packages/financial-templates-lib/src/price-feed/CreatePriceFeed.js#L40

And @mrice32 is right about the interface for `getPriceFeedDecimals`